File: php/src/Mudbase/MudbaseClient.php
<?php

declare(strict_types=1);

namespace App\Mudbase;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\GuzzleException;
use Mudbase\Sdk\Api\AuthenticationApi;
use Mudbase\Sdk\Api\DataApi;
use Mudbase\Sdk\ApiException;
use Mudbase\Sdk\Configuration;
use Mudbase\Sdk\Model\LoginLocalUserRequest;
use Mudbase\Sdk\Model\RefreshTokenRequest;

/**
 * Thin wrapper around the real, generated Mudbase PHP SDK (composer package `mudbase/sdk`,
 * namespace `Mudbase\Sdk`), ported near-verbatim from the sibling social PHP port's client of the
 * same name — see that file for the fuller rationale behind each design choice. This app's
 * surface is intentionally smaller: no registration UI is built (the three demo accounts already
 * exist — see plan/build-plan.md "Auth Model"), so this class drops `registerWithRole()`,
 * `verifyEmail()`, and `establishAnonymousSession()` entirely; it only wraps login/logout/session
 * plus the five collection operations.
 *
 * One real SDK gap this class still works around (documented in README "Known limitations", not a
 * silent workaround): `DataApi::listData()` deserializes into `DataListResponseDataInner`, a typed
 * model that only declares `_id`, `created_at`, `updated_at` — the generator had no way to type a
 * project's user-defined collection fields, and unlike the single-document endpoints (whose `data`
 * property is typed as a passthrough `object`), it did not fall back to a generic map for the list
 * variant. Every actual field (`boardId`, `title`, `position`, ...) is silently dropped by the
 * typed method. This class calls `DataApi::listDataRequest()` — a public method on the generated
 * class that builds the same signed PSR-7 request — and sends it through the SDK's own Guzzle
 * client, then decodes the JSON body itself instead of routing it through the lossy typed
 * deserializer.
 *
 * Access-token refresh: every method that hits the SDK funnels through `callApi()` (ApiException-
 * based calls) or the inline retry in `listDocuments()` (its own raw-Guzzle path, see gap above).
 * Both attempt exactly one silent refresh-and-retry on a 401 using the refresh token this instance
 * was constructed with, via `POST /api/auth/refresh` (`AuthenticationApi::refreshToken()`). A
 * successful refresh calls the optional `$onTokenRefreshed` callback so the caller (bootstrap.php)
 * can persist the new pair into `$_SESSION` — this class has no `$_SESSION` dependency of its own,
 * keeping it usable outside a web request (e.g. from a CLI script). If the refresh itself fails
 * (refresh token expired/revoked too), the original 401 is rethrown as `MudbaseApiError` and the
 * front controller's existing "clear session, redirect to /login" fallback (public/index.php)
 * takes over.
 */
final class MudbaseClient
{
    private readonly Configuration $config;
    private readonly GuzzleClient $httpClient;
    private readonly AuthenticationApi $auth;
    private readonly DataApi $data;

    /**
     * The generated SDK's `DataApi::listDataRequest()`/`listData()` both throw
     * `\InvalidArgumentException` client-side for any `$limit > 100` (confirmed by reading the
     * generated source — `lib/Api/DataApi.php` hard-codes this cap, it is not a server-side
     * constraint this app can query around). `listDocuments()` silently clamps to this rather
     * than letting that exception escape as an uncaught 500 — every caller in this app (board
     * lists/cards, the activity feed) is already operating at demo scale, where 100 is
     * functionally unbounded.
     */
    private const SDK_MAX_LIMIT = 100;

    /**
     * @param (callable(string $token, ?string $refreshToken): void)|null $onTokenRefreshed
     *        Invoked once, synchronously, immediately after a successful silent token refresh —
     *        the caller's chance to persist the new pair (e.g. into `$_SESSION`) before the
     *        retried call goes out with it.
     */
    public function __construct(
        private readonly string $baseUrl,
        private readonly string $projectId,
        private ?string $accessToken = null,
        private ?string $refreshTokenValue = null,
        private readonly mixed $onTokenRefreshed = null,
    ) {
        $this->config = Configuration::getDefaultConfiguration()->setHost($this->baseUrl);
        if ($this->accessToken !== null) {
            $this->config->setAccessToken($this->accessToken);
        }
        $this->httpClient = new GuzzleClient();
        $this->auth = new AuthenticationApi($this->httpClient, $this->config);
        $this->data = new DataApi($this->httpClient, $this->config);
    }

    /**
     * Returns a new client bound to the same project/host but a different bearer token (and,
     * usually, its paired refresh token). Carries the refresh-callback forward so a client handed
     * out after login can still self-heal on a later 401 the same way the original could.
     */
    public function withToken(string $token, ?string $refreshToken = null): self
    {
        return new self($this->baseUrl, $this->projectId, $token, $refreshToken, $this->onTokenRefreshed);
    }

    // ─── Auth ────────────────────────────────────────────────────────────────

    /**
     * @return array{token: string, refreshToken: ?string, user: array<string, mixed>}
     */
    public function login(string $email, string $password): array
    {
        $request = new LoginLocalUserRequest([
            'email' => $email,
            'password' => $password,
            'project_id' => $this->projectId,
        ]);

        $response = $this->callApi(fn () => $this->auth->loginLocalUser($request));

        $token = $response->getToken();
        if ($token === null || $token === '') {
            throw new MudbaseApiError('Login did not return a token', 500);
        }

        $refreshToken = $response->getRefreshToken();
        $authed = $this->withToken($token, $refreshToken);
        return [
            'token' => $token,
            'refreshToken' => $refreshToken,
            'user' => $authed->fetchSessionUser(),
        ];
    }

    /** Best-effort — the caller clears its local session regardless of whether this succeeds. */
    public function logout(): void
    {
        try {
            $this->auth->logoutLocalUser();
        } catch (ApiException) {
            // Session may already be expired/invalid server-side; nothing left to revoke.
        }
    }

    /**
     * `GET /api/auth/local/session` — the authoritative source of the current user's full
     * application state (id, email, name, `customRole`, `isAnonymous`, ...). Deliberately used
     * instead of trusting the narrower typed `user` object returned by login:
     * `GetLocalSession200Response.user` is typed as a passthrough `object` in the OpenAPI spec, so
     * every field the server actually returns survives deserialization, whereas the login
     * response's own typed user object only declares id/email/firstName/lastName/role/
     * emailVerified/twoFactorEnabled — no `customRole`, which `AppContext::isSignedIn()`/`role()`
     * need.
     *
     * @return array<string, mixed>
     */
    public function fetchSessionUser(): array
    {
        $response = $this->callApi(fn () => $this->auth->getLocalSession($this->projectId));

        $user = $response->getUser();
        return is_array($user) ? $user : [];
    }

    // ─── Collections ─────────────────────────────────────────────────────────

    /**
     * @param array<string, mixed>|null $filter
     * @return array{data: list<array<string, mixed>>, pagination: array{page: int, limit: int, total: int, totalPages: int}}
     */
    public function listDocuments(
        string $collectionId,
        ?array $filter = null,
        string $sort = '-createdAt',
        int $page = 1,
        int $limit = 20,
    ): array {
        $filterJson = $filter !== null && $filter !== [] ? json_encode($filter) : null;
        $clampedLimit = min($limit, self::SDK_MAX_LIMIT);

        $status = 500;
        $refreshed = false;
        while (true) {
            try {
                // See class docblock gap — listDataRequest() builds the same signed request the
                // typed listData() would send; we just decode the body ourselves instead of
                // letting it flow through the lossy DataListResponseDataInner model. Rebuilding
                // the request (rather than reusing one PSR-7 instance) on a retry matters: it
                // reads the Bearer header from $this->config fresh each call, so it picks up the
                // token callApi()'s refresh path just wrote into $this->config.
                $request = $this->data->listDataRequest($this->projectId, $collectionId, $page, $clampedLimit, $sort, $filterJson);
                $response = $this->httpClient->send($request, ['http_errors' => false]);
            } catch (GuzzleException $e) {
                throw new MudbaseApiError($e->getMessage(), 502);
            }

            $status = $response->getStatusCode();
            if ($status === 401 && !$refreshed && $this->refreshAccessToken()) {
                $refreshed = true;
                continue;
            }
            break;
        }

        $decoded = json_decode((string) $response->getBody(), true);

        if ($status >= 400) {
            throw MudbaseApiError::fromResponse($status, $decoded);
        }

        $data = is_array($decoded) && is_array($decoded['data'] ?? null) ? $decoded['data'] : [];
        $pagination = is_array($decoded) && is_array($decoded['pagination'] ?? null) ? $decoded['pagination'] : [];

        return [
            'data' => array_values($data),
            'pagination' => [
                'page' => (int) ($pagination['page'] ?? $page),
                'limit' => (int) ($pagination['limit'] ?? $clampedLimit),
                'total' => (int) ($pagination['total'] ?? count($data)),
                'totalPages' => (int) ($pagination['totalPages'] ?? 1),
            ],
        ];
    }

    /** @return array<string, mixed> */
    public function getDocument(string $collectionId, string $documentId): array
    {
        $response = $this->callApi(fn () => $this->data->getData($this->projectId, $collectionId, $documentId));

        $data = $response->getData();
        return is_array($data) ? $data : [];
    }

    /**
     * @param array<string, mixed> $fields
     * @return array<string, mixed>
     */
    public function createDocument(string $collectionId, array $fields): array
    {
        $response = $this->callApi(fn () => $this->data->createData($this->projectId, $collectionId, $fields));

        $data = $response->getData();
        return is_array($data) ? $data : [];
    }

    /**
     * @param array<string, mixed> $fields
     * @return array<string, mixed>
     */
    public function updateDocument(string $collectionId, string $documentId, array $fields): array
    {
        $response = $this->callApi(fn () => $this->data->updateData($this->projectId, $collectionId, $documentId, $fields));

        $data = $response->getData();
        return is_array($data) ? $data : [];
    }

    public function deleteDocument(string $collectionId, string $documentId): void
    {
        $this->callApi(fn () => $this->data->deleteData($this->projectId, $collectionId, $documentId));
    }

    // ─── Token refresh ───────────────────────────────────────────────────────

    /**
     * Runs one SDK call (anything that throws the generated `ApiException` on a non-2xx
     * response). On a 401, attempts exactly one silent refresh via the stored refresh token and,
     * if that succeeds, retries the same call once more with the new token already baked into
     * `$this->config` (every generated Api class shares that same Configuration instance, so the
     * retried call picks it up automatically). Any other status, or a failed/absent refresh,
     * surfaces as the usual `MudbaseApiError`.
     *
     * @template T
     * @param callable(): T $sdkCall
     * @return T
     */
    private function callApi(callable $sdkCall): mixed
    {
        try {
            return $sdkCall();
        } catch (ApiException $e) {
            if ($e->getCode() === 401 && $this->refreshAccessToken()) {
                try {
                    return $sdkCall();
                } catch (ApiException $retryException) {
                    throw MudbaseApiError::fromApiException($retryException);
                }
            }

            throw MudbaseApiError::fromApiException($e);
        }
    }

    /**
     * `POST /api/auth/refresh` — exchanges the stored refresh token for a new access/refresh
     * pair. Returns false (never throws) when there is no refresh token to use or the exchange
     * itself fails, so callers can fall back to surfacing the original 401 instead of masking it
     * with a confusing refresh-endpoint error. On success, mutates this instance's Configuration
     * in place (so already-constructed Api clients pick up the new Bearer token immediately) and
     * invokes `$onTokenRefreshed` so the caller can persist the new pair.
     */
    private function refreshAccessToken(): bool
    {
        if ($this->refreshTokenValue === null || $this->refreshTokenValue === '') {
            return false;
        }

        try {
            $response = $this->auth->refreshToken(new RefreshTokenRequest([
                'refresh_token' => $this->refreshTokenValue,
            ]));
        } catch (ApiException) {
            // Refresh token itself expired/revoked/invalid — nothing left to recover with.
            return false;
        }

        $newToken = $response->getToken();
        if (!is_string($newToken) || $newToken === '') {
            return false;
        }

        $newRefreshToken = $response->getRefreshToken();
        $this->accessToken = $newToken;
        $this->refreshTokenValue = is_string($newRefreshToken) && $newRefreshToken !== ''
            ? $newRefreshToken
            : $this->refreshTokenValue;
        $this->config->setAccessToken($newToken);

        if ($this->onTokenRefreshed !== null) {
            ($this->onTokenRefreshed)($this->accessToken, $this->refreshTokenValue);
        }

        return true;
    }
}

File: python/app/mudbase_client.py
"""Thin synchronous wrapper around the real `mudbase_sdk` (urllib3-based, no
asyncio variant is published), plus small `asyncio.to_thread` adapters so
FastAPI's async handlers never block the event loop on a blocking HTTP call.

Ported verbatim from the sibling `mudbase-showcase-social/python` port (same
SDK version, same platform quirks) — see that project's
`app/mudbase_client.py` docstring for the full history of why `login_sync`
below bypasses the generated typed wrapper method: `login_local_user`'s
generated response's nested user model
(`LoginLocalUser200ResponseUser`) only declares
`id/email/firstName/lastName/role/emailVerified/twoFactorEnabled` — it has
no `customRole` field, even though the real Multi-Role feature returns one.
This app needs `customRole` on every login (it's how owner/member/viewer is
told apart), so the raw response is parsed directly instead of trusting the
incomplete typed model.

This app has no anonymous/guest session and no self-registration UI (see
plan/build-plan.md "Auth Model") — every one of the three roles signs in
with a real account, so there is no `create_anonymous_session_sync` or
`register_with_role_sync` here, unlike the social/ecommerce ports.
"""

import json
from typing import Any

import mudbase_sdk
from mudbase_sdk.rest import ApiException  # type: ignore[attr-defined]  # matches the SDK's own generated doc examples

from app.config import get_settings


class MudbaseApiError(Exception):
    """Raised for any non-2xx Mudbase response, typed or raw."""

    def __init__(self, message: str, status_code: int, code: str | None = None) -> None:
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.code = code


def _build_api_client(access_token: str | None = None) -> mudbase_sdk.ApiClient:
    settings = get_settings()
    configuration = mudbase_sdk.Configuration(host=settings.mudbase_url)
    if access_token:
        configuration.access_token = access_token
    return mudbase_sdk.ApiClient(configuration)


def _raw_json_call(
    api_client: mudbase_sdk.ApiClient,
    method: str,
    resource_path: str,
    path_params: dict[str, str] | None = None,
    body: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Issues a request via the SDK's own public serialize/call primitives and
    parses the JSON body ourselves. See module docstring for why."""
    serialized = api_client.param_serialize(
        method=method,
        resource_path=resource_path,
        path_params=path_params or {},
        query_params=[],
        header_params={"Content-Type": "application/json", "Accept": "application/json"},
        body=body,
        post_params=[],
        files={},
        auth_settings=[],
        collection_formats={},
    )
    response = api_client.call_api(*serialized)
    response.read()  # type: ignore[no-untyped-call]  # mudbase_sdk.rest.RESTResponse.read() ships without a type stub
    raw_text = response.data.decode("utf-8") if response.data else ""
    parsed: dict[str, Any] = json.loads(raw_text) if raw_text else {}

    if not (200 <= response.status < 300):
        message = parsed.get("error") or parsed.get("message") or f"Request failed ({response.status})"
        raise MudbaseApiError(message, response.status, parsed.get("code"))

    return parsed


def _wrap_api_exception(exc: ApiException) -> MudbaseApiError:
    body: dict[str, Any] = {}
    if exc.body:
        try:
            body = json.loads(exc.body)
        except (ValueError, TypeError):
            body = {}
    message = body.get("error") or body.get("message") or exc.reason or "Mudbase request failed"
    return MudbaseApiError(message, exc.status or 500, body.get("code"))


# ─── Auth ───────────────────────────────────────────────────────────────────


def login_sync(email: str, password: str) -> dict[str, Any]:
    settings = get_settings()
    with _build_api_client() as client:
        return _raw_json_call(
            client,
            "POST",
            "/api/auth/local/login",
            body={"email": email, "password": password, "projectId": settings.mudbase_project_id},
        )


def refresh_sync(refresh_token: str) -> dict[str, Any]:
    with _build_api_client() as client:
        return _raw_json_call(
            client,
            "POST",
            "/api/auth/refresh",
            body={"refreshToken": refresh_token},
        )


def logout_sync(access_token: str) -> None:
    with _build_api_client(access_token) as client:
        auth_api = mudbase_sdk.AuthenticationApi(client)
        try:
            auth_api.logout_local_user()
        except ApiException:
            # Best-effort: the caller always clears the local session cookie
            # regardless, so a dead/expired token on logout is not fatal.
            return


# ─── Collections (lists / cards / activity) ─────────────────────────────────


def list_data_sync(
    collection_id: str,
    *,
    filter_dict: dict[str, Any] | None = None,
    sort: str | None = None,
    page: int = 1,
    limit: int = 20,
    access_token: str | None = None,
) -> dict[str, Any]:
    settings = get_settings()
    with _build_api_client(access_token) as client:
        api = mudbase_sdk.DataApi(client)
        filter_str = json.dumps(filter_dict) if filter_dict else None
        try:
            response = api.list_data(
                settings.mudbase_project_id,
                collection_id,
                page=page,
                limit=limit,
                sort=sort,
                filter=filter_str,
            )
        except ApiException as exc:
            raise _wrap_api_exception(exc) from exc
        items = [item.to_dict() for item in (response.data or [])]
        pagination = response.pagination.to_dict() if response.pagination else None
        return {"data": items, "pagination": pagination}


def get_data_sync(collection_id: str, document_id: str, *, access_token: str | None = None) -> dict[str, Any] | None:
    settings = get_settings()
    with _build_api_client(access_token) as client:
        api = mudbase_sdk.DataApi(client)
        try:
            response = api.get_data(settings.mudbase_project_id, collection_id, document_id)
        except ApiException as exc:
            if exc.status == 404:
                return None
            raise _wrap_api_exception(exc) from exc
        return response.data or {}


def create_data_sync(collection_id: str, body: dict[str, Any], *, access_token: str) -> dict[str, Any]:
    settings = get_settings()
    with _build_api_client(access_token) as client:
        api = mudbase_sdk.DataApi(client)
        try:
            response = api.create_data(settings.mudbase_project_id, collection_id, body)
        except ApiException as exc:
            raise _wrap_api_exception(exc) from exc
        return response.data or {}


def update_data_sync(
    collection_id: str,
    document_id: str,
    body: dict[str, Any],
    *,
    access_token: str,
) -> dict[str, Any]:
    settings = get_settings()
    with _build_api_client(access_token) as client:
        api = mudbase_sdk.DataApi(client)
        try:
            response = api.update_data(settings.mudbase_project_id, collection_id, document_id, body)
        except ApiException as exc:
            raise _wrap_api_exception(exc) from exc
        return response.data or {}


def delete_data_sync(collection_id: str, document_id: str, *, access_token: str) -> None:
    settings = get_settings()
    with _build_api_client(access_token) as client:
        api = mudbase_sdk.DataApi(client)
        try:
            api.delete_data(settings.mudbase_project_id, collection_id, document_id)
        except ApiException as exc:
            raise _wrap_api_exception(exc) from exc
